File: go/internal/shared/rebac.go
// Package shared defines the core ReBAC primitives shared by the authz and
// documents services.
//
// These types are the Go equivalents of typescript/src/shared/rebac.ts:
//
//	Object / Relation / Subject  →  named string types (like TS branded types)
//	TupleKey                     →  one edge in the relationship graph
//	CheckRequest / CheckResult   →  the question and answer for a permission check
//
// Both services import this package; neither service owns it.
package shared

import (
	"fmt"
	"strings"
)

// ── Object types ──────────────────────────────────────────────────────────────

// ObjectType is the set of recognised entity kinds in the domain.
type ObjectType string

const (
	ObjectTypeUser      ObjectType = "user"
	ObjectTypeTeam      ObjectType = "team"
	ObjectTypeWorkspace ObjectType = "workspace"
	ObjectTypeDocument  ObjectType = "document"
)

// Object is a fully-qualified entity reference in "type:id" format,
// e.g. "user:alice" or "workspace:productWorkspace".
//
// Using a named string type (rather than plain string) makes it harder to
// accidentally pass a raw string where an Object is expected — the same idea
// as TypeScript's branded string types, enforced by the Go type system.
type Object string

// Relation names a directed edge in the permission graph, e.g. "editor" or "can_read".
type Relation string

const (
	// Team relations
	RelationTeamAdmin  Relation = "admin"
	RelationTeamMember Relation = "member"

	// Workspace relations
	RelationWorkspaceOwner  Relation = "owner"
	RelationWorkspaceEditor Relation = "editor"
	RelationWorkspaceViewer Relation = "viewer"

	// Document structural relation (links a document to its parent workspace)
	RelationDocumentWorkspace Relation = "workspace"

	// Document base relations
	RelationDocumentOwner  Relation = "owner"
	RelationDocumentEditor Relation = "editor"
	RelationDocumentViewer Relation = "viewer"

	// Document computed relations (derived by the graph evaluator)
	RelationDocumentCanRead    Relation = "can_read"
	RelationDocumentCanComment Relation = "can_comment"
	RelationDocumentCanEdit    Relation = "can_edit"
	RelationDocumentCanDelete  Relation = "can_delete"
)

// Subject is either a plain Object ("user:alice") or a subject set
// ("team:platform#member"). IsSubjectSet distinguishes them.
type Subject string

// TupleKey is one edge in the relationship graph:
//
//	(object, relation, user)  →  "team:platform has member user:alice"
type TupleKey struct {
	Object   Object
	Relation Relation
	User     Subject
}

// ── Check types ───────────────────────────────────────────────────────────────

// CheckRequest asks "does User have Relation on Object?"
type CheckRequest struct {
	User     Object
	Relation Relation
	Object   Object
}

// CheckResult is the answer to a CheckRequest.
// Trace is an ordered log of the traversal steps — useful for debugging.
type CheckResult struct {
	Allowed bool
	Trace   []string
}

// ── Constructor helpers ───────────────────────────────────────────────────────

// User returns an Object for a user entity: "user:<id>".
func User(id string) Object { return newObject(ObjectTypeUser, id) }

// Team returns an Object for a team entity: "team:<id>".
func Team(id string) Object { return newObject(ObjectTypeTeam, id) }

// Workspace returns an Object for a workspace entity: "workspace:<id>".
func Workspace(id string) Object { return newObject(ObjectTypeWorkspace, id) }

// Document returns an Object for a document entity: "document:<id>".
func Document(id string) Object { return newObject(ObjectTypeDocument, id) }

// SubjectSet returns a subject-set string like "team:platformTeam#member".
// This is the Go equivalent of the TS subjectSet() helper.
func SubjectSet(obj Object, rel Relation) Subject {
	return Subject(fmt.Sprintf("%s#%s", obj, rel))
}

// Tuple builds a TupleKey. Short-form constructor for use in fixture files.
func Tuple(obj Object, rel Relation, subject Subject) TupleKey {
	return TupleKey{Object: obj, Relation: rel, User: subject}
}

// ── Parsing helpers ───────────────────────────────────────────────────────────

// ParseObject splits "type:id" into its constituent parts.
// Returns an error if the format is invalid or the type is unrecognised.
func ParseObject(s string) (ObjectType, string, error) {
	idx := strings.IndexByte(s, ':')
	if idx < 1 || idx == len(s)-1 {
		return "", "", fmt.Errorf("invalid object %q: want type:id", s)
	}
	typ := ObjectType(s[:idx])
	id := s[idx+1:]
	if !isObjectType(typ) {
		return "", "", fmt.Errorf("unknown object type %q in %q", typ, s)
	}
	return typ, id, nil
}

// ParseSubjectSet splits "team:platformTeam#member" into its object and relation.
func ParseSubjectSet(s Subject) (Object, Relation, error) {
	str := string(s)
	idx := strings.IndexByte(str, '#')
	if idx < 1 || idx == len(str)-1 {
		return "", "", fmt.Errorf("invalid subject set %q: want object#relation", s)
	}
	return Object(str[:idx]), Relation(str[idx+1:]), nil
}

// IsSubjectSet reports whether s is a subject-set reference (contains '#').
func IsSubjectSet(s Subject) bool {
	return strings.ContainsRune(string(s), '#')
}

// ── Private helpers ───────────────────────────────────────────────────────────

func newObject(typ ObjectType, id string) Object {
	if strings.TrimSpace(id) == "" {
		panic(fmt.Sprintf("shared: %s id cannot be empty", typ))
	}
	return Object(fmt.Sprintf("%s:%s", typ, id))
}

func isObjectType(t ObjectType) bool {
	switch t {
	case ObjectTypeUser, ObjectTypeTeam, ObjectTypeWorkspace, ObjectTypeDocument:
		return true
	}
	return false
}

File: typescript/src/shared/rebac.ts
// Shared ReBAC types — the vocabulary that the authz service and any product
// service (documents, billing, etc.) both speak.
//
// In production this would be the published SDK package
// (@your-org/authz-sdk). Here it lives in src/shared/ as a local module.
//
// The authz service owns these concepts; product services import them
// to express check requests and write relationship tuples.

// ── Object types ──────────────────────────────────────────────────────────────

export type ObjectType = "user" | "team" | "workspace" | "document";

// A typed entity reference in "type:id" format.  TypeScript's template literal
// types enforce correct prefixes at compile time.
// Examples: "user:alice", "document:roadmap"
export type RebacObject<TType extends ObjectType = ObjectType> = `${TType}:${string}`;

// ── Relation names ─────────────────────────────────────────────────────────────

export type TeamRelation      = "admin" | "member";
export type WorkspaceRelation = "owner" | "editor" | "viewer";
export type DocumentRelation  =
    | "workspace"
    | "owner" | "editor" | "viewer"
    | "can_read" | "can_comment" | "can_edit" | "can_delete";

export type Relation = TeamRelation | WorkspaceRelation | DocumentRelation;

// ── Subjects ──────────────────────────────────────────────────────────────────

// A subject set: "team:platform#member" = everyone who holds `member` on `team:platform`.
export type TeamSubjectSet = `${RebacObject<"team">}#${TeamRelation}`;
export type SubjectSet     = TeamSubjectSet;
export type Subject        = RebacObject | SubjectSet;

// ── Tuple ─────────────────────────────────────────────────────────────────────

// One edge in the relationship graph.  Asserts "`user` has `relation` on `object`".
// Field named `user` to match OpenFGA's tuple API — it can be a concrete object
// or a subject set.
export type TupleKey = {
    object:   RebacObject;
    relation: Relation;
    user:     Subject;
};

// ── Check types ───────────────────────────────────────────────────────────────

export type CheckRequest = {
    user:     RebacObject<"user">;
    relation: Relation;
    object:   RebacObject;
};

export type CheckResult = {
    allowed: boolean;
    trace:   string[];
};

// ── Helper constructors ───────────────────────────────────────────────────────

export const user      = (id: string): RebacObject<"user">      => makeObject("user", id);
export const team      = (id: string): RebacObject<"team">      => makeObject("team", id);
export const workspace = (id: string): RebacObject<"workspace"> => makeObject("workspace", id);
export const document  = (id: string): RebacObject<"document">  => makeObject("document", id);

export const subjectSet = (
    objectId: RebacObject<"team">,
    relation: TeamRelation,
): TeamSubjectSet => `${objectId}#${relation}`;

export const tuple = (
    objectId: RebacObject,
    relation: Relation,
    subject:  Subject,
): TupleKey => ({ object: objectId, relation, user: subject });

// ── Parse helpers ─────────────────────────────────────────────────────────────

export const parseObject = (value: string): { type: ObjectType; id: string } => {
    const [type, ...rest] = value.split(":");
    const id = rest.join(":");
    if (!isObjectType(type) || id.length === 0) {
        throw new Error(`Invalid ReBAC object: ${value}`);
    }
    return { type, id };
};

export const parseSubjectSet = (
    value: string,
): { object: RebacObject<"team">; relation: TeamRelation } => {
    const [objectId, relation] = value.split("#");
    if (!objectId || !isObjectOfType(objectId, "team") || !isTeamRelation(relation)) {
        throw new Error(`Invalid subject set: ${value}`);
    }
    return { object: objectId, relation };
};

export const isSubjectSet = (subject: Subject): subject is SubjectSet =>
    subject.includes("#");

export const isObjectOfType = <T extends ObjectType>(
    value: string,
    type: T,
): value is RebacObject<T> => {
    try { return parseObject(value).type === type; }
    catch { return false; }
};

// ── Private ───────────────────────────────────────────────────────────────────

const makeObject = <T extends ObjectType>(type: T, id: string): RebacObject<T> => {
    if (id.trim().length === 0) throw new Error(`${type} id cannot be empty`);
    return `${type}:${id}`;
};

const isObjectType = (v: string | undefined): v is ObjectType =>
    v === "user" || v === "team" || v === "workspace" || v === "document";

const isTeamRelation = (v: string | undefined): v is TeamRelation =>
    v === "admin" || v === "member";
